### PR TITLE
Step 4 (updated) and Step 5

### PR DESCRIPTION
**Pull request by shinjatta**
## What
1. Added db file
2. Implemented prepareDB() function
3. Implemented dbData() that gets data from the database
4. Changed addItem() so that it does an insert to the database
5. Deleted jsonData() and "encoding/json" from the inserts because they were no longer in use
6. Added the database and the json file in the .gitignore
7. Added dockerFile
8.  Implemented an addCategory() function in case the user wants to add an item to a non existing category
9. Changed the getItem() function so it just gets one item

## CHECKS :warning:

Please make sure you are aware of the following.

- [ ] **The changes in this PR doesn't have private information

### CODE REVIEW

File: go/app/main.go [hunk 1]
```diff
@@ -35,36 +36,63 @@ type Response struct {
...
 
-func root(c echo.Context) error {
-	res := Response{Message: "Hello, world!"}
-	return c.JSON(http.StatusOK, res)
+func prepareDB() {
+	database, err := sql.Open("sqlite3", "mercari.db")
+	if err != nil {
+		log.Fatal(err)
+	}
+	statement, _ := database.Prepare(`
```
Comment (tsbkw) [inline go/app/main.go]: Maybe it is better to log this error (`_`) to get a more straightforward error message.

File: go/app/main.go [hunk 2]
```diff
@@ -79,33 +107,43 @@ func addItem(c echo.Context) error {
...
+	prepareDB()
+	database, err := sql.Open("sqlite3", "mercari.db")
 	if err != nil {
-		fmt.Println(err)
+		log.Fatal(err)
+	}
+
+	//Insert the data into the database
+	statement, _ := database.Prepare("INSERT INTO `Items` (`id`, `name`, `category_id`, `image_filename`) VALUES (?, ?, ?, ?);")
+	if err != nil {
```
Comment (tsbkw) [inline go/app/main.go:133]: This err is assigned at L118. Maybe your intention is change `_` to err in L124.

File: go/app/main.go [hunk 3]
```diff
@@ -79,33 +107,43 @@ func addItem(c echo.Context) error {
...
+	err = database.QueryRow("SELECT id FROM Items ORDER BY id DESC LIMIT 1").Scan(&itemID)
+	if err != nil {
+		log.Fatal(err)
+	}
 
-	//Save into a JSON file
-	content, err := json.Marshal(items)
+	//Getting the id corresponding to the category that was given
+	var categoryID int
+	err = database.QueryRow("SELECT id FROM Category WHERE name = ?", category).Scan(&categoryID)
```
Comment (tsbkw) [inline go/app/main.go:140]: In current implementation, there is no way to create new category. Could you try to create new category if not found?

File: go/app/main.go [hunk 4]
```diff
@@ -79,33 +107,43 @@ func addItem(c echo.Context) error {
...
+	defer statement.Close()
 
-	//Add new 
-	items.Items = append(items.Items, newItem)
+	//Getting the id corresponding to the last item
+	var itemID int
+	err = database.QueryRow("SELECT id FROM Items ORDER BY id DESC LIMIT 1").Scan(&itemID)
+	if err != nil {
+		log.Fatal(err)
+	}
```
Comment (tsbkw) [inline go/app/main.go]: OPTIONAL:
FYI, some database has feature `autoincrement` and it is convenient for creating unique integer id.
In sqlite, if you create table with `INTEGER PRIMARY KEY`, and omit `id` from insertion statement, you can use this feature.
see 1. Summary of https://www.sqlite.org/autoinc.html

File: go/app/main.go [hunk 5]
```diff
@@ -129,15 +167,16 @@ func getImg(c echo.Context) error {
...
+	items, err := dbData()
 	if err != nil {
 		fmt.Println(err)
 	}
 	return c.JSON(http.StatusOK, items)
 }
 
 func getItem(c echo.Context) error {
-	items, err := dataJson()
+	items, err := dbData()
```
Comment (tsbkw) [inline go/app/main.go]: OPTIONAL:
In json, there is no way to fetch specific item_id record, however, you can fetch specific record by SQL.
If you have time, could you try get only item_id specified, instead of fetching all data?
You can do it with another PR and only when you have time (maybe finish other steps)

**Review by tsbkw — COMMENTED** [review]
I commented what I noticed. OPTIONAL comments are just for your information, and can be don when you have time :bow:

OPTIONAL: if you have time after finishing other steps, could you try to handle errors during API request and return error instead of stopping application?
In production code, it is better to use `log.Fatal` only when application should be stopped, and other cases we should handle error with logging, and return errors to client.

File: go/app/main.go (hunk 5 above)
Comment (shinjatta) [reply go/app/main.go]: Yeah, I don't know how I did not directly did the SQL query to get the item. Solved. I guess I was just being lazy and reusing too much code. But much better this way, true.

File: go/Dockerfile [hunk 6]
```diff
@@ -1,8 +1,26 @@
-FROM alpine
+FROM golang:latest
 
-RUN addgroup -S mercari && adduser -S trainee -G mercari
-# RUN chown -R trainee:mercari /path/to/db
+# Set the working directory inside the container
+WORKDIR /go/app
 
-USER trainee
```
Comment (tsbkw) [inline go/Dockerfile]: If there is no reason, could you preserve using USER trainee?

see: https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#user

File: go/app/main.go [hunk 7]
```diff
@@ -121,36 +126,55 @@ func addItem(c echo.Context) error {
...
 	var categoryID int
 	err = database.QueryRow("SELECT id FROM Category WHERE name = ?", category).Scan(&categoryID)
 	if err != nil {
 		fmt.Println("This category does not exist")
-		log.Fatal(err)
+		addCategory(category)
+		err = database.QueryRow("SELECT id FROM Category WHERE name = ?", category).Scan(&categoryID)
+		if err != nil {
+			log.Fatal(err)
+		}
```
Comment (tsbkw) [inline go/app/main.go]: Could you try to use sql.Result in L172?
This query checking id can be replaced with sql.Result.LastInsertId in addCategory and addCategory can return inserted category's id.
https://pkg.go.dev/database/sql#Result

File: go/app/main.go [hunk 8]
```diff
@@ -121,36 +126,55 @@ func addItem(c echo.Context) error {
...
 		log.Fatal(err)
 	}
 
 	return c.JSON(http.StatusOK, res)
 }
 
+//addCategory is called when there is no Category when creating a new item
+func addCategory(category string) {
+	prepareDB()
+	database, err := sql.Open("sqlite3", "mercari.db")
```
Comment (tsbkw) [inline go/app/main.go:162]: I overlooked database is not closed. Could you close databases like statements?
https://pkg.go.dev/database/sql#DB.Close

File: go/app/main.go [hunk 9]
```diff
@@ -55,8 +55,12 @@ func prepareDB() {
 	);
 	`)
 	statement.Exec()
+	if err != nil {
+		log.Fatal(err)
+	}
```
Comment (tsbkw) [inline go/app/main.go:59]: This err should be checked before Exec() is called.
Also, statement.Exec may return error, so it should be checked.
https://pkg.go.dev/database/sql#Stmt.Exec

**Review by tsbkw — COMMENTED** [review]
Could you check comments when you have time? :bow:

This time you can continue but it is recommended to create another PR for each Step (or any other meaningful chunk).

File: .github/workflows/build.yml [hunk 10]
```diff
@@ -37,10 +37,10 @@ jobs:
...
-#      with:
-#       context: <go or python>
-#       push: true
-#       tags: ${{ steps.meta.outputs.tags }}
-#       labels: ${{ steps.meta.outputs.labels }}
\ No newline at end of file
+    - name: Build and push Docker image
+      uses: docker/build-push-action@ad44023a93711e3deb337508980b4b5e9bcdc5dc
+      with:
+        context: <go or python>
```
Comment (tsbkw) [inline .github/workflows/build.yml]: Here, you have to choose build context. You choose go for develop, so specify folder `go` like below.
```suggestion
        context: go
```

Document: https://github.com/docker/build-push-action#customizing

**Review by tsbkw — COMMENTED** [review]
Could you change as I commented and check no error occurs in github actions.
